### PR TITLE
Implementation of Light bulb

### PR DESCRIPTION
**Pull request by jbsky**
This contribution makes it possible to operate the lb1 class. It is based on bg1 by adapting the fields specific to lb1.
I have updated the comments so that other developers can be guided.

The starting point was #434

### CODE REVIEW

**Comment by felipediel** [comment]
Looks very good. Two questions.

1. Is it possible to use set_state() to define `bulb_scenes`?
2. What happened to `bulb_sceneidx`?

**Comment by jbsky** [comment]
1/ Not with this code
2/ I don't know. To be honest, I haven't tested bulb_scenes

If you are objective, with the code that currently exists, I doubt very much that you can drive a light bulb with it.

Before implementing bulb_scenes, you need to know how it works. Maybe a comment could have helped. Could you at least test bulb_scenes with the existing one?

Maybe I'll implement later.

File: broadlink/light.py [hunk 1]
```diff
@@ -9,66 +10,91 @@
...
+        pwr: bool = None,
+        red: int = None,
+        blue: int = None,
+        green: int = None,
+        brightness: int = None,
+        colortemp: int = None,
+        hue: int = None,
+        saturation: int = None,
+        transitionduration: int = None,
+        maxworktime: int = None,
```
Comment (felipediel) [inline broadlink/light.py:39]: What happens when we define `maxworktime`?

**Comment by felipediel** [comment]
Hi @jbsky. I made some improvements and implemented the missing features. Mind helping with tests? I don't have the device.

**Comment by jbsky** [comment]
Hi felipediel,

Given the lack of testing, it goes fast, there is no regression.

maxworktime is the time in seconds before the lamp turns off.

bulbscene, please, for later 😃

**Comment by felipediel** [comment]
Let's at least see if that effect dict works.
```python3
import broadlink as blk
d = blk.hello("192.168.0.16")   # Example device.
d.auth()
```
We can start having a look at the state.
```python3
print(d.get_state())
```
 Then we set an effect.
```python3
effects = {
    "lovely color": 0,
    "flashlight": 1,
    "lightning": 2,
    "color fading": 3,
    "color breathing": 4,
    "multicolor breathing": 5,
    "color jumping": 6,
    "multicolor jumping": 7,
}
d.set_state(bulb_sceneidx=effects["color_fading"])
print(d.get_state())
```
We can also try an invalid index to see what happens:
```python3
d.set_state(bulb_sceneidx=50)
print(d.get_state())
```
255 probably means no effect.
```python3
d.set_state(bulb_sceneidx=255)
print(d.get_state())
```

**Comment by jbsky** [comment]
When a bulb_sceneidx is correctly set, => 'bulb_colormode' pass to value 2.


d.set_state(bulb_sceneidx=-1]) => nothing effect
d.set_state(bulb_sceneidx=0) => work
d.set_state(bulb_sceneidx=1) => work
...
d.set_state(bulb_sceneidx=7) => work
d.set_state(bulb_sceneidx=0) =>  nothing effect
d.set_state(bulb_sceneidx=10) => nothing effect
d.set_state(bulb_sceneidx=256) => nothing effect

This array seems to have the correctly named effects.

**Comment by felipediel** [comment]
I am working on an update. In the example you put in the code, the bulb_sceneidx is 255. Is it correct to interpret 255 as "no scene"?

**Comment by jbsky** [comment]
The bulb has 3 modes you can set
bulb_colormode=0 => rgb => bulb_sceneidx is set to 255
bulb_colormode=1 => no color => bulb_sceneidx is set to 255
bulb_colormode=2 => scene, => bulb_sceneidx is set to 0 as we just tested, the scenes can be triggered either by providing id from 0 to 7 for bulb_sceneidx.

You can't set 255 directly to bulb_sceneidx.

**Comment by felipediel** [comment]
Thank you very much! Just a few more questions...

About `bulb_scenes`...
1. Are they related to `bulb_sceneidx`?
2. Is it possible that, by adding more scenes, `bulb_sceneidx` is greater than 7?
3. Is it possible that, by setting `bulb_scenes` to a different color or timing, `bulb_sceneidx` will change the effect corresponding to that index?
4. Do you think you can get a sample of a `bulb_scene` value?
```

**Comment by felipediel** [comment]
I was thinking on something like this:
```python3
class lb1(device):
    """Controls a Broadlink LB1."""

    @enum.unique
    class ColorMode(enum.IntEnum):
        """Enumerates color modes."""
        RGB = 0
        WHITE = 1
        SCENE = 2

    @enum.unique
    class Scene(enum.IntEnum):
        """Enumerates bulb scenes."""
        LOVELY_COLOR = 0
        FLASHLIGHT = 1
        LIGHTNING = 2
        COLOR_FADING = 3
        COLOR_BREATHING = 4
        MULTICOLOR_BREATHING = 5
        COLOR_JUMPING = 6
        MULTICOLOR_JUMPING = 7
        NO_SCENE = 255
```
With a permissive interface:
```python3
    def set_state(
        self,
        pwr: bool = None,
        red: int = None,
        blue: int = None,
        green: int = None,
        brightness: int = None,
        colortemp: int = None,
        hue: int = None,
        saturation: int = None,
        transitionduration: int = None,
        maxworktime: int = None,
        bulb_colormode: lb1.ColorMode = None,
        bulb_scenes: str = None,
        bulb_scene: str = None,
        bulb_sceneidx: typing.Union[lb1.Scene, int] = None,
    ) -> dict:
```

But if the scenes are mutable, it makes more sense to get rid of the effects dict and allow the client applications to name those indexes as they wish.

**Comment by jbsky** [comment]
For class scene => NO_SCENE = 255 work. I set bulb_sceneidx to 255 and the scene stop where it is.


I make some test to find how to work bulb_scene only with #RRGGBB format.
it seems to follow this operation : 
"#RRGGBB,sustain time, release time"
sustain time € [1;+∞]
release time € [0;+∞]
If you put a 0 after a release time, the scene stop and loop to begin (good!)
If you put 0 after the first color, you break bulb_scenes >  7. (bad)

```python
>>> green_release = "#000000,1000,0,#00FF00,1,3000,0"
>>> scenes.append(green_release)
>>> d.set_state(bulb_scenes=json.dumps(scenes))
{'red': 166, 'blue': 5, 'green': 5, 'pwr': 1, 'brightness': 100, 'colortemp': 2700, 'hue': 0, 'saturation': 97, 'transitionduration': 1500, 'maxworktime': 0, 'bulb_colormode': 2, 'bulb_scenes': '["@01686464,0,0,0", "#ffffff,10,0,#000000,190,0,0", "2700+100,0,0,0", "#ff0000,500,2500,#00FF00,500,2500,#0000FF,500,2500,0", "@01686464,100,2400,@01686401,100,2400,0", "@01686464,100,2400,@01686401,100,2400,@005a6464,100,2400,@005a6401,100,2400,0", "@01686464,10,0,@00000000,190,0,0", "@01686464,200,0,@005a6464,200,0,0", "#000000,1000,0,#0080FF,1000,5000,0", "#000000,1000,0,#00FF00,1,3000,0"]', 'bulb_scene': '#000000,1000,0,#0080FF,1000,5000,0', 'bulb_sceneidx': 8}
>>> d.set_state(bulb_sceneidx=len(scenes)-1)
{'red': 166, 'blue': 5, 'green': 5, 'pwr': 1, 'brightness': 100, 'colortemp': 2700, 'hue': 0, 'saturation': 97, 'transitionduration': 1500, 'maxworktime': 0, 'bulb_colormode': 2, 'bulb_scenes': '["@01686464,0,0,0", "#ffffff,10,0,#000000,190,0,0", "2700+100,0,0,0", "#ff0000,500,2500,#00FF00,500,2500,#0000FF,500,2500,0", "@01686464,100,2400,@01686401,100,2400,0", "@01686464,100,2400,@01686401,100,2400,@005a6464,100,2400,@005a6401,100,2400,0", "@01686464,10,0,@00000000,190,0,0", "@01686464,200,0,@005a6464,200,0,0", "#000000,1000,0,#0080FF,1000,5000,0", "#000000,1000,0,#00FF00,1,3000,0"]', 'bulb_scene': '#000000,1000,0,#00FF00,1,3000,0', 'bulb_sceneidx': 9}
```

**Comment by felipediel** [comment]
Great! I think we already have the interface and the information we need to implement these features in client applications. We are good to go. Thank you very much!

**Comment by jbsky** [comment]
Hi Felipediel, 

I wanted to know when the next release is scheduled.

I would like to switch to Home Assistant and it looks like it is based on releases. Being a beginner on HA, I'd like to make things easier.

Thank you in advance,

Yours sincerely,

Julien

**Comment by kylemd** [comment]
> Hi Felipediel,
> 
> I wanted to know when the next release is scheduled.
> 
> I would like to switch to Home Assistant and it looks like it is based on releases. Being a beginner on HA, I'd like to make things easier.
> 
> Thank you in advance,
> 
> Yours sincerely,
> 
> Julien

Per [this comment](https://github.com/mjg59/python-broadlink/pull/540#issuecomment-787694965) the functionality for light bulbs hasn't been developed for Home Assistant yet although it now works from command line.

Furthermore, it will not automatically update in Home Assistant once it does, as the original developer (@mjg59) needs to set up PyPi for it to publish (as referenced [here](https://github.com/mjg59/python-broadlink/issues/480#issuecomment-784711740))

Long story short you'll be waiting a while - but good things come to those who wait ;)

If it is implemented you would be able to force installing it until its' been added officially per [this comment](https://github.com/mjg59/python-broadlink/issues/480#issuecomment-752183830)
